packages/vitest/src/node/plugins/index.ts

```diff
@@ -64,12 +64,15 @@ export async function VitestPlugin(options: UserConfig = {}, ctx = new Vitest('t
 
         const config: ViteConfig = {
           root: viteConfig.test?.root || options.root,
-          esbuild: {
-            sourcemap: 'external',
-
-            // Enables using ignore hint for coverage providers with @preserve keyword
-            legalComments: 'inline',
-          },
+          esbuild: viteConfig.esbuild === false
+            ? false
+            : {
+                ...viteConfig.esbuild,
```

Comment:
You don't need to merge it, the result of this function is merged with the custom config